File: correct_coordinates_crawlee.js
/**
 * Correct coordinates for all locations using Crawlee + Puppeteer to scrape Google Maps
 * This script scrapes coordinates directly from Google Maps without using the paid API.
 */

const { PuppeteerCrawler } = require('crawlee');
const fs = require('fs').promises;
const path = require('path');

const PROJECT_ROOT = path.resolve(__dirname, '..');
const ENRICHED_FILE = path.join(PROJECT_ROOT, 'scraped_data', 'enriched_spots.json');
const BACKUP_FILE = path.join(PROJECT_ROOT, 'scraped_data', 'enriched_spots.json.backup');
const CORRECTED_FILE = path.join(PROJECT_ROOT, 'scraped_data', 'enriched_spots_corrected.json');

// Distance threshold in kilometers
const DISTANCE_THRESHOLD_KM = 5.0;
const REQUEST_DELAY = 500; // 0.5 seconds (reduced for speed)
const MAX_CONCURRENCY = 3; // Process 3 locations in parallel
const PAGE_WAIT_TIME = 1500; // 1.5 seconds (reduced from 3s)

/**
 * Calculate distance between two coordinates using Haversine formula
 */
function calculateDistance(lat1, lon1, lat2, lon2) {
    const R = 6371; // Earth's radius in kilometers
    const dLat = toRad(lat2 - lat1);
    const dLon = toRad(lon2 - lon1);
    
    const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
              Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) *
              Math.sin(dLon / 2) * Math.sin(dLon / 2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    
    return R * c;
}

function toRad(degrees) {
    return degrees * (Math.PI / 180);
}

/**
 * Parse coordinate string in DMS format (e.g., "2°28'47.0"N 103°15'48.5"E")
 */
function parseDMSCoordinates(coordString) {
    // Pattern: degrees°minutes'seconds"direction (e.g., 2°28'47.0"N 103°15'48.5"E)
    const dmsPattern = /(\d+)°(\d+)'([\d.]+)"([NS])\s+(\d+)°(\d+)'([\d.]+)"([EW])/;
    const match = coordString.match(dmsPattern);
    
    if (match) {
        const latDeg = parseFloat(match[1]);
        const latMin = parseFloat(match[2]);
        const latSec = parseFloat(match[3]);
        const latDir = match[4];
        
        const lngDeg = parseFloat(match[5]);
        const lngMin = parseFloat(match[6]);
        const lngSec = parseFloat(match[7]);
        const lngDir = match[8];
        
        // Convert DMS to decimal
        let lat = latDeg + latMin / 60 + latSec / 3600;
        let lng = lngDeg + lngMin / 60 + lngSec / 3600;
        
        // Apply direction
        if (latDir === 'S') lat = -lat;
        if (lngDir === 'W') lng = -lng;
        
        // Validate coordinates (Malaysia is roughly 0-7°N, 99-120°E)
        if (lat >= 0 && lat <= 10 && lng >= 99 && lng <= 120) {
            return [lat, lng];
        }
    }
    
    return null;
}

/**
 * Extract coordinates from Google Maps URL
 */
function extractCoordsFromUrl(url) {
    // Pattern: @lat,lng or /@lat,lng,zoom
    const patterns = [
        /@(-?\d+\.?\d*),(-?\d+\.?\d*)/,  // @lat,lng
        /\/@(-?\d+\.?\d*),(-?\d+\.?\d*)/,  // /@lat,lng
        /!3d(-?\d+\.?\d*)!4d(-?\d+\.?\d*)/,  // !3dlat!4dlng (old format)
    ];
    
    for (const pattern of patterns) {
        const match = url.match(pattern);
        if (match) {
            try {
                const lat = parseFloat(match[1]);
                const lng = parseFloat(match[2]);
                // Validate coordinates (Malaysia is roughly 0-7°N, 99-120°E)
                if (lat >= 0 && lat <= 10 && lng >= 99 && lng <= 120) {
                    return [lat, lng];
                }
            } catch (e) {
                continue;
            }
        }
    }
    
    return null;
}

/**
 * Scrape coordinates from Google Maps using Crawlee PuppeteerCrawler
 * Optimized version with batch processing
 */
async function scrapeGoogleMapsCoordsBatch(requests) {
    const results = new Map();
    
    const crawler = new PuppeteerCrawler({
        launchContext: {
            launchOptions: {
                headless: true,
                args: ['--no-sandbox', '--disable-setuid-sandbox'], // Faster startup
            },
        },
        maxConcurrency: MAX_CONCURRENCY,
        requestHandler: async ({ request, page }) => {
            try {
                // Wait for initial page load
                await page.waitForNavigation({ waitUntil: 'networkidle2', timeout: 15000 }).catch(() => {});
                await new Promise(resolve => setTimeout(resolve, 2000));
                
                let currentUrl = page.url();
                
                // Debug: Log URL for first few requests
                if (request.userData.id <= 3) {
                    console.log(`    🔍 Initial URL: ${currentUrl.substring(0, 100)}...`);
                }
                
                // Check if we're on a search results page (not a specific location)
                // Google Maps search results have "/search/" in URL but no coordinates
                const isSearchResultsPage = currentUrl.includes('/search/') && !extractCoordsFromUrl(currentUrl);
                
                if (isSearchResultsPage) {
                    // Click on the first search result to get to the location page
                    try {
                        // Try multiple selectors for the first result
                        const firstResultSelectors = [
                            'a[href*="/place/"]',
                            'div[role="main"] a[data-value*="@"]',
                            'div[jsaction*="click"] a[href*="/place/"]',
                            'div[data-result-index="0"] a',
                            'div.m6QErb a[href*="/place/"]'
                        ];
                        
                        let clicked = false;
                        for (const selector of firstResultSelectors) {
                            try {
                                await page.waitForSelector(selector, { timeout: 3000 });
                                const firstResult = await page.$(selector);
                                if (firstResult) {
                                    await firstResult.click();
                                    clicked = true;
                                    break;
                                }
                            } catch (e) {
                                continue;
                            }
                        }
                        
                        if (clicked) {
                            // Wait for navigation to location page
                            await page.waitForNavigation({ waitUntil: 'networkidle2', timeout: 10000 }).catch(() => {});
                            await new Promise(resolve => setTimeout(resolve, 2000));
                            currentUrl = page.url();
                            
                            if (request.userData.id <= 3) {
                                console.log(`    ✅ Clicked first result, new URL: ${currentUrl.substring(0, 100)}...`);
                            }
                        }
                    } catch (e) {
                        if (request.userData.id <= 3) {
                            console.log(`    ⚠️  Could not click first result: ${e.message}`);
                        }
                    }
                }
                
                // Try to extract coordinates from URL
                let coords = extractCoordsFromUrl(currentUrl);
                
                if (coords) {
                    if (request.userData.id <= 3) {
                        console.log(`    ✅ Found coords in URL: ${coords[0]}, ${coords[1]}`);
                    }
                    results.set(request.userData.id, coords);
                    return;
                }
                
                // Try alternative: Look for data in the page HTML
                try {
                    // Wait a bit more for map to fully load
                    await new Promise(resolve => setTimeout(resolve, 1000));
                    
                    // Get page content to search for coordinates
                    const pageContent = await page.content();
                    
                    // Look for coordinates in the page HTML (Google Maps embeds them in various formats)
                    const coordPatterns = [
                        /@(-?\d+\.?\d*),(-?\d+\.?\d*)/g,
                        /!3d(-?\d+\.?\d*)!4d(-?\d+\.?\d*)/g,
                        /"lat":\s*(-?\d+\.?\d*).*?"lng":\s*(-?\d+\.?\d*)/g,
                        /\[(-?\d+\.?\d*),\s*(-?\d+\.?\d*)\]/g,
                        /center.*?\[(-?\d+\.?\d*),\s*(-?\d+\.?\d*)\]/g,
                        /data-value="[^"]*@(-?\d+\.?\d*),(-?\d+\.?\d*)/g
                    ];
                    
                    for (const pattern of coordPatterns) {
                        const matches = [...pageContent.matchAll(pattern)];
                        for (const match of matches) {
                            const lat = parseFloat(match[1]);
                            const lng = parseFloat(match[2]);
                            if (lat >= 0 && lat <= 10 && lng >= 99 && lng <= 120) {
                                coords = [lat, lng];
                                if (request.userData.id <= 3) {
                                    console.log(`    ✅ Found coords in HTML: ${lat}, ${lng}`);
                                }
                                break;
                            }
                        }
                        if (coords) break;
                    }
                    
                    if (coords) {
                        results.set(request.userData.id, coords);
                        return;
                    }
                    
                } catch (e) {
                    // Ignore extraction errors
                }
                
                // If still no coords, log for debugging
                if (request.userData.id <= 3) {
                    console.log(`    ⚠️  No coordinates found in URL or page content`);
                }
                
                results.set(request.userData.id, null);
            } catch (error) {
                console.log(`    ⚠️  Error scraping ${request.userData.placeName || 'location'}: ${error.message}`);
                results.set(request.userData.id, null);
            }
        },
    });
    
    // Run crawler for all requests
    await crawler.run(requests);
    
    return results;
}

async function scrapeGoogleMapsCoords(placeName, address = '') {
    const query = placeName || address;
    if (!query) {
        return null;
    }
    
    // Add "Malaysia" to help with search accuracy
    const searchQuery = query.includes('Malaysia') || address.includes('Malaysia') 
        ? query 
        : `${query}, Malaysia`;
    
    const searchUrl = `https://www.google.com/maps/search/${encodeURIComponent(searchQuery)}`;
    
    const request = { 
        url: searchUrl, 
        userData: { 
            id: placeName,
            name: placeName,
            result: null 
        } 
    };
    
    const results = await scrapeGoogleMapsCoordsBatch([request]);
    return results.get(placeName) || null;
}

/**
 * Main function to correct coordinates
 */
async function correctCoordinates() {
    console.log('='.repeat(80));
    console.log('CORRECTING COORDINATES USING GOOGLE MAPS SCRAPING (CRAWLEE/PUPPETEER)');
    console.log('='.repeat(80));
    
    // Load enriched spots data
    let enrichedData;
    try {
        const data = await fs.readFile(ENRICHED_FILE, 'utf-8');
        enrichedData = JSON.parse(data);
        console.log(`\n📂 Loaded ${enrichedData.length} locations from: ${ENRICHED_FILE}`);
    } catch (error) {
        console.error(`❌ Error loading file: ${error.message}`);
        return;
    }
    
    // Create backup
    console.log(`\n💾 Creating backup: ${BACKUP_FILE}`);
    await fs.writeFile(BACKUP_FILE, JSON.stringify(enrichedData, null, 2), 'utf-8');
    console.log('✅ Backup created');
    
    // Process each location
    let correctedCount = 0;
    let unchangedCount = 0;
    let errorCount = 0;
    const corrections = [];
    
    console.log(`\n🔍 Processing locations (threshold: ${DISTANCE_THRESHOLD_KM}km, concurrency: ${MAX_CONCURRENCY})...`);
    console.log('-'.repeat(80));
    
    // Filter locations that need processing
    const locationsToProcess = enrichedData
        .map((location, idx) => {
            const googleData = location.google_maps_data || {};
            const placeName = googleData.place_name || '';
            const address = googleData.address || '';
            const currentLat = location.latitude || 0;
            const currentLng = location.longitude || 0;
            
            // Skip if no place name or invalid coordinates
            if (!placeName || currentLat === 0 || currentLng === 0) {
                return null;
            }
            
            return { idx, location, placeName, address, currentLat, currentLng };
        })
        .filter(item => item !== null);
    
    console.log(`📊 Processing ${locationsToProcess.length} locations (skipping ${enrichedData.length - locationsToProcess.length} with missing data)\n`);
    
    // Process in batches for better performance
    const batchSize = MAX_CONCURRENCY;
    for (let i = 0; i < locationsToProcess.length; i += batchSize) {
        const batch = locationsToProcess.slice(i, i + batchSize);
        
        // Check if place names are already coordinates (DMS format)
        const batchWithCoords = batch.map(({ idx, placeName, address, currentLat, currentLng, location }) => {
            // Check if place name is already in coordinate format
            const dmsCoords = parseDMSCoordinates(placeName);
            if (dmsCoords) {
                return {
                    idx,
                    location,
                    placeName,
                    address,
                    currentLat,
                    currentLng,
                    preParsedCoords: dmsCoords
                };
            }
            return {
                idx,
                location,
                placeName,
                address,
                currentLat,
                currentLng,
                preParsedCoords: null
            };
        });
        
        // Prepare batch requests (only for locations that need scraping)
        const batchRequests = batchWithCoords
            .filter(item => !item.preParsedCoords)
            .map(({ idx, placeName, address, currentLat, currentLng, location }) => {
                const query = placeName || address;
                const searchQuery = query.includes('Malaysia') || address.includes('Malaysia') 
                    ? query 
                    : `${query}, Malaysia`;
                const searchUrl = `https://www.google.com/maps/search/${encodeURIComponent(searchQuery)}`;
                
                return {
                    url: searchUrl,
                    userData: {
                        id: idx,
                        idx,
                        placeName,
                        address,
                        currentLat,
                        currentLng,
                        location
                    }
                };
            });
        
        // Log batch start
        batchWithCoords.forEach(({ idx, placeName, preParsedCoords }) => {
            if (preParsedCoords) {
                console.log(`[${idx + 1}/${enrichedData.length}] ${placeName} (DMS format - parsing directly)`);
            } else {
                console.log(`[${idx + 1}/${enrichedData.length}] ${placeName}`);
            }
        });
        
        // Scrape all in batch (only for non-DMS locations)
        const results = batchRequests.length > 0 
            ? await scrapeGoogleMapsCoordsBatch(batchRequests)
            : new Map();
        
        // Add pre-parsed DMS coordinates to results
        batchWithCoords.forEach(({ idx, preParsedCoords }) => {
            if (preParsedCoords) {
                results.set(idx, preParsedCoords);
            }
        });
        
        // Process results
        for (const { idx, location, placeName, currentLat, currentLng, preParsedCoords } of batchWithCoords) {
            const newCoords = results.get(idx);
            
            if (newCoords) {
                const [newLat, newLng] = newCoords;
                
                // If this was pre-parsed from DMS, show it
                if (preParsedCoords) {
                    console.log(`  ✅ Parsed from DMS: (${newLat.toFixed(6)}, ${newLng.toFixed(6)})`);
                }
                
                const distance = calculateDistance(currentLat, currentLng, newLat, newLng);
                
                if (distance > DISTANCE_THRESHOLD_KM) {
                    console.log(`  📍 Current: (${currentLat.toFixed(6)}, ${currentLng.toFixed(6)})`);
                    console.log(`  ✅ Corrected: (${newLat.toFixed(6)}, ${newLng.toFixed(6)})`);
                    console.log(`  📏 Distance: ${distance.toFixed(2)} km`);
                    
                    // Update coordinates
                    location.latitude = newLat;
                    location.longitude = newLng;
                    correctedCount++;
                    
                    corrections.push({
                        index: idx,
                        name: placeName,
                        oldCoords: [currentLat, currentLng],
                        newCoords: [newLat, newLng],
                        distanceKm: distance
                    });
                } else {
                    console.log(`  ✓ Coordinates OK (difference: ${distance.toFixed(2)} km)`);
                    unchangedCount++;
                }
            } else {
                console.log(`  ⚠️  Could not scrape coordinates, keeping original`);
                errorCount++;
            }
        }
        
        // Progress update
        const processed = Math.min(i + batchSize, locationsToProcess.length);
        console.log(`\n📊 Progress: ${processed}/${locationsToProcess.length} processed (${correctedCount} corrected, ${unchangedCount} OK, ${errorCount} errors)\n`);
        
        // Small delay between batches
        if (i + batchSize < locationsToProcess.length) {
            await new Promise(resolve => setTimeout(resolve, REQUEST_DELAY));
        }
    }
    
    // Count skipped locations
    unchangedCount += (enrichedData.length - locationsToProcess.length);
    
    // Save corrected data
    console.log(`\n${'='.repeat(80)}`);
    console.log('SUMMARY');
    console.log('='.repeat(80));
    console.log(`Total locations: ${enrichedData.length}`);
    console.log(`✅ Corrected: ${correctedCount}`);
    console.log(`✓ Unchanged: ${unchangedCount}`);
    console.log(`⚠️  Errors: ${errorCount}`);
    
    if (correctedCount > 0) {
        console.log(`\n💾 Saving corrected data to: ${CORRECTED_FILE}`);
        await fs.writeFile(CORRECTED_FILE, JSON.stringify(enrichedData, null, 2), 'utf-8');
        console.log('✅ Corrected data saved');
        
        // Also update the original file
        console.log(`\n💾 Updating original file: ${ENRICHED_FILE}`);
        await fs.writeFile(ENRICHED_FILE, JSON.stringify(enrichedData, null, 2), 'utf-8');
        console.log('✅ Original file updated');
        
        // Show corrections summary
        console.log(`\n📋 Corrections made:`);
        console.log('-'.repeat(80));
        corrections.slice(0, 10).forEach(correction => {
            console.log(`  [${correction.index}] ${correction.name}`);
            console.log(`      Old: ${correction.oldCoords[0].toFixed(6)}, ${correction.oldCoords[1].toFixed(6)}`);
            console.log(`      New: ${correction.newCoords[0].toFixed(6)}, ${correction.newCoords[1].toFixed(6)}`);
            console.log(`      Distance: ${correction.distanceKm.toFixed(2)} km`);
        });
        if (corrections.length > 10) {
            console.log(`  ... and ${corrections.length - 10} more corrections`);
        }
    } else {
        console.log('\n✅ No corrections needed - all coordinates are accurate!');
    }
    
    console.log('='.repeat(80));
    console.log(`\n💡 Next steps:`);
    console.log(`   1. Review the corrected data in: ${CORRECTED_FILE}`);
    console.log(`   2. If satisfied, the original file has been updated`);
    console.log(`   3. Run fix_spots_mapping_v2.py to regenerate spots-simple.json`);
    console.log('='.repeat(80));
}

// Run the script
if (require.main === module) {
    correctCoordinates().catch(console.error);
}

module.exports = { correctCoordinates, scrapeGoogleMapsCoords };

File: correct_coordinates_crawlee.py
#!/usr/bin/env python3
"""
Correct coordinates for all locations using Crawlee + Puppeteer to scrape Google Maps
This script scrapes coordinates directly from Google Maps without using the paid API.
"""

import json
import time
import re
import asyncio
from pathlib import Path
from typing import Dict, Optional, Tuple
import math

# Note: This script uses Playwright (via Crawlee) which is more reliable than Puppeteer
# For Puppeteer, you'd need to use crawlee-puppeteer-crawler, but Playwright is recommended

MVP_DIR = Path(__file__).parent
PROJECT_ROOT = MVP_DIR.parent
ENRICHED_FILE = PROJECT_ROOT / 'scraped_data' / 'enriched_spots.json'
BACKUP_FILE = PROJECT_ROOT / 'scraped_data' / 'enriched_spots.json.backup'
CORRECTED_FILE = PROJECT_ROOT / 'scraped_data' / 'enriched_spots_corrected.json'

# Distance threshold in kilometers - if coordinates differ by more than this, update them
DISTANCE_THRESHOLD_KM = 5.0  # 5km threshold

# Delay between requests to avoid being blocked
REQUEST_DELAY = 2.0  # 2 seconds between requests


def calculate_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculate distance between two coordinates in kilometers using Haversine formula."""
    R = 6371  # Earth's radius in kilometers
    
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    
    a = (math.sin(dlat / 2) ** 2 +
         math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
         math.sin(dlon / 2) ** 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    
    return R * c


def extract_coords_from_url(url: str) -> Optional[Tuple[float, float]]:
    """Extract coordinates from Google Maps URL."""
    # Pattern: @lat,lng or /@lat,lng,zoom
    patterns = [
        r'@(-?\d+\.?\d*),(-?\d+\.?\d*)',  # @lat,lng
        r'/@(-?\d+\.?\d*),(-?\d+\.?\d*)',  # /@lat,lng
        r'!3d(-?\d+\.?\d*)!4d(-?\d+\.?\d*)',  # !3dlat!4dlng (old format)
    ]
    
    for pattern in patterns:
        match = re.search(pattern, url)
        if match:
            try:
                lat = float(match.group(1))
                lng = float(match.group(2))
                # Validate coordinates (Malaysia is roughly 0-7°N, 99-120°E)
                if 0 <= lat <= 10 and 99 <= lng <= 120:
                    return (lat, lng)
            except ValueError:
                continue
    
    return None


async def scrape_google_maps_coords(place_name: str, address: str = "") -> Optional[Tuple[float, float]]:
    """
    Scrape coordinates from Google Maps using Playwright (via Crawlee).
    Returns (latitude, longitude) or None if not found.
    """
    try:
        from playwright.async_api import async_playwright
        
        # Build search query
        query = place_name if place_name else address
        if not query:
            return None
        
        # Add "Malaysia" to help with search accuracy
        if "Malaysia" not in query and "Malaysia" not in address:
            query = f"{query}, Malaysia"
        
        async with async_playwright() as p:
            # Launch browser (headless by default)
            browser = await p.chromium.launch(headless=True)
            context = await browser.create_context(
                user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
            )
            page = await context.new_page()
            
            try:
                # Navigate to Google Maps search
                search_url = f"https://www.google.com/maps/search/{query.replace(' ', '+')}"
                await page.goto(search_url, wait_until='networkidle', timeout=30000)
                
                # Wait for map to load
                await page.wait_for_timeout(3000)
                
                # Try to get coordinates from URL (most reliable)
                current_url = page.url
                coords = extract_coords_from_url(current_url)
                
                if coords:
                    await browser.close()
                    return coords
                
                # Alternative: Try to extract from page content
                # Look for data attributes or script tags with coordinates
                try:
                    # Check if there's a data attribute with coordinates
                    coords_element = await page.query_selector('[data-value*="@"]')
                    if coords_element:
                        data_value = await coords_element.get_attribute('data-value')
                        if data_value:
                            coords = extract_coords_from_url(data_value)
                            if coords:
                                await browser.close()
                                return coords
                except:
                    pass
                
                # Try to get from page title or meta tags
                try:
                    # Sometimes coordinates are in the page title
                    title = await page.title()
                    coords = extract_coords_from_url(title)
                    if coords:
                        await browser.close()
                        return coords
                except:
                    pass
                
                await browser.close()
                return None
                
            except Exception as e:
                print(f"    ⚠️  Error scraping: {str(e)}")
                await browser.close()
                return None
                
    except ImportError:
        print("\n❌ ERROR: Playwright not installed!")
        print("\nTo install:")
        print("  pip install playwright")
        print("  playwright install chromium")
        return None
    except Exception as e:
        print(f"    ❌ Error: {str(e)}")
        return None


def correct_coordinates():
    """Main function to correct coordinates for all locations."""
    print("="*80)
    print("CORRECTING COORDINATES USING GOOGLE MAPS SCRAPING (CRAWLEE/PLAYWRIGHT)")
    print("="*80)
    
    # Load enriched spots data
    if not ENRICHED_FILE.exists():
        print(f"❌ Error: {ENRICHED_FILE} not found!")
        return
    
    print(f"\n📂 Loading data from: {ENRICHED_FILE}")
    with open(ENRICHED_FILE, 'r', encoding='utf-8') as f:
        enriched_data = json.load(f)
    
    print(f"✅ Loaded {len(enriched_data)} locations")
    
    # Create backup
    print(f"\n💾 Creating backup: {BACKUP_FILE}")
    with open(BACKUP_FILE, 'w', encoding='utf-8') as f:
        json.dump(enriched_data, f, indent=2, ensure_ascii=False)
    print("✅ Backup created")
    
    # Process each location
    corrected_count = 0
    unchanged_count = 0
    error_count = 0
    corrections = []
    
    print(f"\n🔍 Processing locations (threshold: {DISTANCE_THRESHOLD_KM}km)...")
    print("-" * 80)
    
    async def process_locations():
        nonlocal corrected_count, unchanged_count, error_count, corrections
        
        for idx, location in enumerate(enriched_data):
            google_data = location.get('google_maps_data', {})
            place_name = google_data.get('place_name', '')
            address = google_data.get('address', '')
            current_lat = location.get('latitude', 0)
            current_lng = location.get('longitude', 0)
            
            # Skip if no place name
            if not place_name:
                unchanged_count += 1
                continue
            
            # Scrape coordinates
            print(f"\n[{idx + 1}/{len(enriched_data)}] {place_name}")
            new_coords = await scrape_google_maps_coords(place_name, address)
            
            # Rate limiting
            await asyncio.sleep(REQUEST_DELAY)
            
            if new_coords:
                new_lat, new_lng = new_coords
                distance = calculate_distance(current_lat, current_lng, new_lat, new_lng)
                
                if distance > DISTANCE_THRESHOLD_KM:
                    print(f"  📍 Current: ({current_lat:.6f}, {current_lng:.6f})")
                    print(f"  ✅ Corrected: ({new_lat:.6f}, {new_lng:.6f})")
                    print(f"  📏 Distance: {distance:.2f} km")
                    
                    # Update coordinates
                    location['latitude'] = new_lat
                    location['longitude'] = new_lng
                    corrected_count += 1
                    
                    corrections.append({
                        'index': idx,
                        'name': place_name,
                        'old_coords': (current_lat, current_lng),
                        'new_coords': (new_lat, new_lng),
                        'distance_km': distance
                    })
                else:
                    print(f"  ✓ Coordinates OK (difference: {distance:.2f} km)")
                    unchanged_count += 1
            else:
                print(f"  ⚠️  Could not scrape coordinates, keeping original")
                error_count += 1
    
    # Run async processing
    asyncio.run(process_locations())
    
    # Save corrected data
    print(f"\n" + "="*80)
    print("SUMMARY")
    print("="*80)
    print(f"Total locations: {len(enriched_data)}")
    print(f"✅ Corrected: {corrected_count}")
    print(f"✓ Unchanged: {unchanged_count}")
    print(f"⚠️  Errors: {error_count}")
    
    if corrected_count > 0:
        print(f"\n💾 Saving corrected data to: {CORRECTED_FILE}")
        with open(CORRECTED_FILE, 'w', encoding='utf-8') as f:
            json.dump(enriched_data, f, indent=2, ensure_ascii=False)
        print("✅ Corrected data saved")
        
        # Also update the original file
        print(f"\n💾 Updating original file: {ENRICHED_FILE}")
        with open(ENRICHED_FILE, 'w', encoding='utf-8') as f:
            json.dump(enriched_data, f, indent=2, ensure_ascii=False)
        print("✅ Original file updated")
        
        # Show corrections summary
        print(f"\n📋 Corrections made:")
        print("-" * 80)
        for correction in corrections[:10]:  # Show first 10
            print(f"  [{correction['index']}] {correction['name']}")
            print(f"      Old: {correction['old_coords'][0]:.6f}, {correction['old_coords'][1]:.6f}")
            print(f"      New: {correction['new_coords'][0]:.6f}, {correction['new_coords'][1]:.6f}")
            print(f"      Distance: {correction['distance_km']:.2f} km")
        if len(corrections) > 10:
            print(f"  ... and {len(corrections) - 10} more corrections")
    else:
        print("\n✅ No corrections needed - all coordinates are accurate!")
    
    print("="*80)
    print(f"\n💡 Next steps:")
    print(f"   1. Review the corrected data in: {CORRECTED_FILE}")
    print(f"   2. If satisfied, the original file has been updated")
    print(f"   3. Run fix_spots_mapping_v2.py to regenerate spots-simple.json")
    print("="*80)


if __name__ == '__main__':
    correct_coordinates()

